### PR TITLE
Core/Player: remaining XP overflow correction

### PR DESCRIPTION
**Pull request by matanshukry**
**Changes proposed:**

- Fix an issue with exp overflow when the experience required for next level is less than the experience you have (how is that even possible, I'm not really sure)

**Target branch(es):** 3.3.5/master

- [ ] 3.3.5
- [x] master

**Issues addressed:**

**Tests performed:**

(Does it build, tested in-game, etc.)

- [x] It builds
- [] tested in-game
   - I don't know how to get to a situation where the player has more experience than required for that level. Even at the max level the required experience is somewhat high, and the experience is set to 0.

**Known issues and TODO list:** (add/remove lines as needed)

### CODE REVIEW

**Comment by matanshukry** [comment]
@Aokromes comments? specifically, how do you get a character to have more xp than xp required?

**Comment by Aokromes** [comment]
have level 80 (on 3.3.5a branch)

**Comment by matanshukry** [comment]
Doesn't work on master. Not 60 (max level) nor 80.

**Comment by Aokromes** [comment]
well, i did the comment on 3.3.5a branch, not on master branch.

**Comment by matanshukry** [comment]
@Aokromes I wasn't complaining about you or anything like that, just wrote that to clarify I can't fully test it on master.

**Comment by illfated** [comment]
title suggestion: **Core/Player: remaining XP overflow correction**
(or what you prefer, just to follow title codestyle)

**Comment by Aokromes** [comment]
> @Aokromes I wasn't complaining about you or anything like that, just wrote that to clarify I can't fully test it on master.

i mean, i only have seen the bug on 3.3.5a branch, idk on master branch because i don't use it, every level 80 char shows that huge amount of xp left.

File: src/server/game/Handlers/CharacterHandler.cpp [hunk 1]
```diff
@@ -1353,7 +1353,7 @@ void WorldSession::HandlePlayerLogin(LoginQueryHolder* holder)
 
     TC_LOG_INFO("entities.player.character", "Account: %u (IP: %s) Login Character: [%s] (%s) Level: %d, XP: %u/%u (%u left)",
         GetAccountId(), GetRemoteAddress().c_str(), pCurrChar->GetName().c_str(), pCurrChar->GetGUID().ToString().c_str(), pCurrChar->getLevel(),
-        _player->GetXP(), _player->GetXPForNextLevel(), _player->GetXPForNextLevel() - _player->GetXP());
+        _player->GetXP(), _player->GetXPForNextLevel(), std::max(0, (int)_player->GetXPForNextLevel() - (int)_player->GetXP()));
```
Comment (Shauren) [inline src/server/game/Handlers/CharacterHandler.cpp:1356]: cast to int32 (never use raw built in int types)

**Comment by Aokromes** [comment]
Level: 80, XP: 57387/0 (4294909909 left)
 Level: 80, XP: 632/0 (4294966664 left)
 Level: 80, XP: 901/0 (4294966395 left)

**Comment by jackpoz** [comment]
@matanshukry on 335 GetXPForNextLevel() returns 0 when maxed, that's why the difference is negative. In a way checking if GetXPForNextLevel() is not 0 should be enough, something like
```cpp
_player->GetXPForNextLevel() ? _player->GetXPForNextLevel() - _player->GetXP() : 0
```

**Comment by Shauren** [comment]
@jackpoz this version works for all branches and is best to keep that way to have same code everywhere possible

(as for why it behaves differently on master - xp per level is now found in client files and is provided for all levels, even beyond the cap)

**Review by Shauren — APPROVED** [review]

**Pull request merged**